pkg/kubelet/cri/remote/conversion_test.go

```diff
@@ -0,0 +1,319 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remote
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/stretchr/testify/assert"
+	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
+	"k8s.io/cri-api/pkg/apis/runtime/v1alpha2"
+)
+
+var (
+	testLabels      = map[string]string{"labelKey": "labelValue", "foo": "bar"}
+	testAnnotations = map[string]string{"annotationKey": "annotationValue", "foo": "bar"}
+)
+
+func fillFields(s interface{}) {
+	fillFieldsOffset(s, 0)
+}
+
+func fillFieldsOffset(s interface{}, offset int) {
+	reflectType := reflect.TypeOf(s).Elem()
+	reflectValue := reflect.ValueOf(s).Elem()
+
+	for i := 0; i < reflectType.NumField(); i++ {
+		field := reflectValue.Field(i)
+		typeName := reflectType.Field(i).Name
+
+		// Skipping protobuf internal values
+		if strings.HasPrefix(typeName, "XXX_") {
+			continue
+		}
+
+		fillField(field, i+offset)
+	}
+}
+
+func fillField(field reflect.Value, v int) {
+	switch field.Kind() {
+	case reflect.Bool:
+		field.SetBool(true)
+
+	case reflect.Float32, reflect.Float64:
+		field.SetFloat(float64(v))
+
+	case reflect.String:
+		field.SetString(fmt.Sprint(v))
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		field.SetInt(int64(v))
+
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64:
+		field.SetUint(uint64(v))
+
+	case reflect.Map:
+		field.Set(reflect.MakeMap(field.Type()))
+
+	case reflect.Array, reflect.Slice:
+		slice := reflect.MakeSlice(field.Type(), 1, 1)
+		field.Set(slice)
+		first := slice.Index(0)
+
+		if first.Type().Kind() == reflect.Ptr {
+			first.Set(reflect.New(first.Type().Elem()))
+			fillFieldsOffset(first.Interface(), v)
+		} else {
+			fillField(first, v)
+		}
+
+	case reflect.Ptr:
+		val := reflect.New(field.Type().Elem())
+		field.Set(val)
+		fillFieldsOffset(field.Interface(), v)
+
+	case reflect.Struct:
+		fillFieldsOffset(field.Addr().Interface(), v)
+	}
+
+}
+
+func assertEqual(t *testing.T, a, b proto.Message) {
+	aBytes, err := proto.Marshal(a)
+	assert.Nil(t, err)
+
+	bBytes, err := proto.Marshal(b)
+	assert.Nil(t, err)
+
+	assert.Equal(t, aBytes, bBytes)
+}
```

Comment:
@liggitt I decided to go this simpler path for comparing the bytes, since we also skip setting the `XXX_` proto fields in `fillFieldsOffset`. WDYT?

Reply:
The wire serialization isn't what we're concerned about comparing here, we actually need to compare the go type memory equivalence. Copying the function I linked to is probably the simplest way to do that quickly

Reply:
@liggitt okay, but what's the easiest way to get `*types.Type` from any other type?

Reply:
oh, hmm... I didn't realize that wasn't walking reflect types :-(

I don't know if it's easier to convert the equality check to be in terms of reflect types, or use the gengo Builder + FindTypes() call